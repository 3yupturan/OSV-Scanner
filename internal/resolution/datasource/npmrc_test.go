package datasource_test

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"testing"

	"github.com/google/osv-scanner/internal/resolution/datasource"
	"github.com/google/osv-scanner/internal/testutility"
)

// These tests rely on using 'globalconfig' and 'userconfig' in the package .npmrc to override their default locations.
// It's also possible for environment variables or the builtin npmrc to mess with these tests.
// TODO: Should test the default and other methods of setting the global/user config.

func createTempNpmrc(t *testing.T, filename string) string {
	t.Helper()
	dir := testutility.CreateTestDir(t)
	file := filepath.Join(dir, filename)
	f, err := os.Create(file)
	if err != nil {
		t.Fatalf("could not create test npmrc file: %v", err)
	}
	f.Close()

	return file
}

func writeToNpmrc(t *testing.T, file string, lines ...string) {
	t.Helper()
	f, err := os.OpenFile(file, os.O_APPEND|os.O_WRONLY, 0666)
	if err != nil {
		t.Fatalf("could not write to test npmrc file: %v", err)
	}
	defer f.Close()
	for _, line := range lines {
		if _, err := fmt.Fprintln(f, line); err != nil {
			t.Fatalf("could not write to test npmrc file: %v", err)
		}
	}
}

type testNpmrcFiles struct {
	global  string
	user    string
	project string
}

func makeBlankNpmrcFiles(t *testing.T) testNpmrcFiles {
	t.Helper()
	var files testNpmrcFiles
	files.global = createTempNpmrc(t, "npmrc")
	files.user = createTempNpmrc(t, ".npmrc")
	files.project = createTempNpmrc(t, ".npmrc")
	writeToNpmrc(t, files.project, "globalconfig="+files.global, "userconfig="+files.user)

	return files
}

func checkNpmRegistryRequest(t *testing.T, config datasource.NpmRegistryConfig, urlComponents ...string) {
	t.Helper()
	req, err := config.BuildRequest(context.Background(), urlComponents...)
	if err != nil {
		t.Fatalf("error building request: %v", err)
	}

	snapshot := testutility.NewSnapshot()
	snapshot.MatchText(t, req.URL.String())
	snapshot.MatchJSON(t, req.Header["Authorization"])
}

func TestNpmrcNoRegistries(t *testing.T) {
	t.Parallel()
	npmrcFiles := makeBlankNpmrcFiles(t)

	config, err := datasource.LoadNpmRegistryConfig(filepath.Dir(npmrcFiles.project))
	if err != nil {
		t.Fatalf("could not parse npmrc: %v", err)
	}

	if nRegs := len(config); nRegs != 1 {
		t.Errorf("expected 1 npm registry, got %v", nRegs)
	}

	checkNpmRegistryRequest(t, config, "@test/package", "1.2.3")
}

func TestNpmrcRegistryAuth(t *testing.T) {
	t.Parallel()
	npmrcFiles := makeBlankNpmrcFiles(t)
	writeToNpmrc(t, npmrcFiles.project,
		"registry=https://registry1.test.com",
		"//registry1.test.com/:_auth=bXVjaDphdXRoCg==",
		"@test1:registry=https://registry2.test.com",
		"//registry2.test.com/:_authToken=c3VjaCB0b2tlbgo=",
		"@test2:registry=https://sub.registry2.test.com",
		"//sub.registry2.test.com:username=user",
		"//sub.registry2.test.com:_password=d293Cg==",
	)

	config, err := datasource.LoadNpmRegistryConfig(filepath.Dir(npmrcFiles.project))
	if err != nil {
		t.Fatalf("could not parse npmrc: %v", err)
	}

	checkNpmRegistryRequest(t, config, "foo")
	checkNpmRegistryRequest(t, config, "@test0/bar")
	checkNpmRegistryRequest(t, config, "@test1/baz")
	checkNpmRegistryRequest(t, config, "@test2/test")
}

// Do not make this test parallel because it calls t.Setenv()
func TestNpmrcRegistryOverriding(t *testing.T) {
	check := func(t *testing.T, npmrcFiles testNpmrcFiles) {
		t.Helper()
		config, err := datasource.LoadNpmRegistryConfig(filepath.Dir(npmrcFiles.project))
		if err != nil {
			t.Fatalf("could not parse npmrc: %v", err)
		}
		checkNpmRegistryRequest(t, config, "pkg")
		checkNpmRegistryRequest(t, config, "@general/pkg")
		checkNpmRegistryRequest(t, config, "@global/pkg")
		checkNpmRegistryRequest(t, config, "@user/pkg")
		checkNpmRegistryRequest(t, config, "@project/pkg")
	}

	npmrcFiles := makeBlankNpmrcFiles(t)
	writeToNpmrc(t, npmrcFiles.project, "@project:registry=https://project.registry.com")
	writeToNpmrc(t, npmrcFiles.user, "@user:registry=https://user.registry.com")
	writeToNpmrc(t, npmrcFiles.global,
		"@global:registry=https://global.registry.com",
		"@general:registry=https://general.global.registry.com",
		"registry=https://global.registry.com",
	)
	check(t, npmrcFiles)

	// override global in user
	writeToNpmrc(t, npmrcFiles.user,
		"@general:registry=https://general.user.registry.com",
		"registry=https://user.registry.com",
	)
	check(t, npmrcFiles)

	// override global/user in project
	writeToNpmrc(t, npmrcFiles.project,
		"@general:registry=https://general.project.registry.com",
		"registry=https://project.registry.com",
	)
	check(t, npmrcFiles)

	// override global/user/project in environment variable
	t.Setenv("NPM_CONFIG_REGISTRY", "https://environ.registry.com")
	check(t, npmrcFiles)
}
