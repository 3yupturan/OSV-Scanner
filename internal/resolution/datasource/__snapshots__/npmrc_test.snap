
[TestNpmrcRegistryOverriding - 1]
https://global.registry.com/pkg
---

[TestNpmrcRegistryOverriding - 2]
null
---

[TestNpmrcRegistryOverriding - 3]
https://general.global.registry.com/@general%2Fpkg
---

[TestNpmrcRegistryOverriding - 4]
null
---

[TestNpmrcRegistryOverriding - 5]
https://global.registry.com/@global%2Fpkg
---

[TestNpmrcRegistryOverriding - 6]
null
---

[TestNpmrcRegistryOverriding - 7]
https://user.registry.com/@user%2Fpkg
---

[TestNpmrcRegistryOverriding - 8]
null
---

[TestNpmrcRegistryOverriding - 9]
https://project.registry.com/@project%2Fpkg
---

[TestNpmrcRegistryOverriding - 10]
null
---

[TestNpmrcRegistryOverriding - 11]
https://user.registry.com/pkg
---

[TestNpmrcRegistryOverriding - 12]
null
---

[TestNpmrcRegistryOverriding - 13]
https://general.user.registry.com/@general%2Fpkg
---

[TestNpmrcRegistryOverriding - 14]
null
---

[TestNpmrcRegistryOverriding - 15]
https://global.registry.com/@global%2Fpkg
---

[TestNpmrcRegistryOverriding - 16]
null
---

[TestNpmrcRegistryOverriding - 17]
https://user.registry.com/@user%2Fpkg
---

[TestNpmrcRegistryOverriding - 18]
null
---

[TestNpmrcRegistryOverriding - 19]
https://project.registry.com/@project%2Fpkg
---

[TestNpmrcRegistryOverriding - 20]
null
---

[TestNpmrcRegistryOverriding - 21]
https://project.registry.com/pkg
---

[TestNpmrcRegistryOverriding - 22]
null
---

[TestNpmrcRegistryOverriding - 23]
https://general.project.registry.com/@general%2Fpkg
---

[TestNpmrcRegistryOverriding - 24]
null
---

[TestNpmrcRegistryOverriding - 25]
https://global.registry.com/@global%2Fpkg
---

[TestNpmrcRegistryOverriding - 26]
null
---

[TestNpmrcRegistryOverriding - 27]
https://user.registry.com/@user%2Fpkg
---

[TestNpmrcRegistryOverriding - 28]
null
---

[TestNpmrcRegistryOverriding - 29]
https://project.registry.com/@project%2Fpkg
---

[TestNpmrcRegistryOverriding - 30]
null
---

[TestNpmrcRegistryOverriding - 31]
https://environ.registry.com/pkg
---

[TestNpmrcRegistryOverriding - 32]
null
---

[TestNpmrcRegistryOverriding - 33]
https://general.project.registry.com/@general%2Fpkg
---

[TestNpmrcRegistryOverriding - 34]
null
---

[TestNpmrcRegistryOverriding - 35]
https://global.registry.com/@global%2Fpkg
---

[TestNpmrcRegistryOverriding - 36]
null
---

[TestNpmrcRegistryOverriding - 37]
https://user.registry.com/@user%2Fpkg
---

[TestNpmrcRegistryOverriding - 38]
null
---

[TestNpmrcRegistryOverriding - 39]
https://project.registry.com/@project%2Fpkg
---

[TestNpmrcRegistryOverriding - 40]
null
---

[TestNpmrcNoRegistries - 1]
https://registry.npmjs.org/@test%2Fpackage/1.2.3
---

[TestNpmrcNoRegistries - 2]
null
---

[TestNpmrcRegistryAuth - 1]
https://registry1.test.com/foo
---

[TestNpmrcRegistryAuth - 2]
[
  "Basic bXVjaDphdXRoCg=="
]
---

[TestNpmrcRegistryAuth - 3]
https://registry1.test.com/@test0%2Fbar
---

[TestNpmrcRegistryAuth - 4]
[
  "Basic bXVjaDphdXRoCg=="
]
---

[TestNpmrcRegistryAuth - 5]
https://registry2.test.com/@test1%2Fbaz
---

[TestNpmrcRegistryAuth - 6]
[
  "Bearer c3VjaCB0b2tlbgo="
]
---

[TestNpmrcRegistryAuth - 7]
https://sub.registry2.test.com/@test2%2Ftest
---

[TestNpmrcRegistryAuth - 8]
[
  "Basic dXNlcjp3b3cK"
]
---
